test/smt/smt-upgrade.test.ts

```diff
@@ -0,0 +1,153 @@
+      poseidon3Elements.address,
+      "SmtV2Mock"
+    );
+
+    // upgrade smt library version
+    const StateV2Factory = await ethers.getContractFactory("StateV2", {
+      libraries: { Smt: smtV2Lib.address },
+    });
+
+    // upgrade state to new version
+    const newState = await upgrades.upgradeProxy(
+      state.address,
+      StateV2Factory,
+      {
+        unsafeAllowLinkedLibraries: true,
+      }
+    );
+    console.log("upgradeProxy tx", newState.address);
+    await newState.deployed();
+
+    expect(state.address).to.equal(newState.address);
+    expect(await newState.getSmtRootHistoryLength()).to.equal(1);
+
+    const { id: id2 } = await publishState(newState, issuerStateTransitions[1]);
+    const history2 = await newState.getSmtRootHistory(1, 1);
+    const proof2 = await state.getSmtProof(id2);
+
+    expect(proof2.root).to.equal(history2[0].RootHistory);
+    expect(await newState.getSmtRootHistoryLength()).to.equal(2);
+  });
+
+  it("library upgrade and state upgrade to a new version with extending functionality", async () => {
+    // publish state to state v2
+    await publishInitialState(state, issuerStateTransitions[0]);
+
+    //deploy new smt library version
+    const smtV2Lib = await stateDeployHelper.deploySmt(
+      poseidon2Elements.address,
+      poseidon3Elements.address,
+      "SmtV2Mock"
+    );
+
+    // upgrade smt library version
+    const StateV2MockFactory = await ethers.getContractFactory("StateV2Mock", {
+      libraries: { SmtV2Mock: smtV2Lib.address },
+    });
+
+    // upgrade state to new version
+    const newState = await upgrades.upgradeProxy(
+      state.address,
+      StateV2MockFactory,
+      {
+        unsafeAllowLinkedLibraries: true,
+      }
+    );
+    console.log("upgradeProxy tx", newState.address);
+    await newState.deployed();
+
+    expect(state.address).to.equal(newState.address);
+    expect(await newState.getSmtRootHistoryLength()).to.equal(1);
+
+    const { id: id2 } = await publishState(newState, issuerStateTransitions[1]);
+    const history2 = await newState.getSmtRootHistory(0, 1);
+    const proof2 = await newState.getSmtProof(id2);
+
+    expect(proof2.root).to.equal(history2[1].RootHistory);
+    expect(proof2.test).to.equal(123);
+    expect(await newState.getSmtRootHistoryLength()).to.equal(2);
+    expect((await newState.rootHistoryLast()).RootHistory).to.equal(
+      history2[1].RootHistory
+    );
+  });
+
+  it("library struct modification to struct and state upgrade", async () => {
+    // publish state to state v2
+    await publishInitialState(state, issuerStateTransitions[0]);
+
+    //deploy new smt library version
+    const smtV2Lib = await stateDeployHelper.deploySmt(
+      poseidon2Elements.address,
+      poseidon3Elements.address,
+      "SmtV3Mock"
+    );
+
+    // upgrade smt library version
+    const StateV3MockFactory = await ethers.getContractFactory("StateV3Mock", {
+      libraries: { SmtV3Mock: smtV2Lib.address },
```

Comment:
It's hard mentally to follow when mixing different versions like `const smtV2Lib` and `SmtV3Mock`

Reply:
fixed